doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,726 @@
+query graph is built in a bottom-up manner, we can use the input quantifiers for resolving names within the
+current part of the query being processed.
+
+It is important to restate the constraint mentioned above: all column references in expressions within each box
+*must* only point to quantifiers either within the same box or within an ancestor box through a chain of unique
+children (correlation).
+
+This section explains how the query graph model being built can be used for name resolution purposes using the following
+query as an example:
+
+![Name resolution](qgm/name-resolution-1.svg)
+
+When planning the `WHERE` clause, belongs in box 0 as shown above, the resulting expression must only reference
+quantifiers Q0 and Q5, however the relations visible in the scope according to the SQL standard are the relations
+represented by Q0, Q1 and Q4, ie. the leaf quantifiers of the comma join (represented by box 0).
+
+A new `NameResolutionContext` struct will encapsulate the name resolution logic, which resolves column names
+against the leaf quantifiers but lifts the column references through the projection of the intermediate boxes,
+all the way up to the current box.
+
+Following with the example, when resolving the reference to `y` in the `WHERE` clause, we will find that among
+the leaf quantifiers (Q0, Q1, Q4), only quantifier Q4 projects a column named `y`, so the column is resolved as
+`Q4.c0`, ie. the first column among the columns projected by Q4's input box (box 4).
+Since we must return a expression referencing only Q0 and Q5, we need to follow the linked chain made by
+`Quantifier::parent_box` and `QueryBox::ranging_quantifiers` until we reach a quantifier ranged over by
+box 0. The parent box of `Q4` is box 2, which projects `Q4.c0` as its forth column and it's ranged over by Q5.
+Therefore, following that chain, we have resolved that `y` means `Q5.c3` within the context of box 0.
+
+In the query above had and explicit projection or an ORDER BY clause names would be resolved against the same
+leaf quantifiers and the resulting column references lifted following the same process.
+
+Basically, a `NameResolutionContext` instance will represent the context of the `FROM` clause.
+
+#### Name resolution within subqueries
+
+Expressions within subqueries must see the symbols visible from the context the subquery is in. To support that
+`NameResultionContext` will contain an optional reference to a parent `NameResolutionContext`, that is passed
+down for planning the subquery, so that if a name cannot be resolved against the context of the `FROM` clause of
+the subquery, we go through the chain of parent contexts until we find a symbol that matches in one of them.
+
+#### Name resolution within the `ON` and `USING` clauses
+
+In the following example, the binary `LEFT JOIN` is represented by box 2, and hence, the `ON` clause belongs in
+that box.
+
+![Name resolution within the `ON` clause](qgm/name-resolution-3.svg)
```

Comment:
In this example, when trying to resolve `f1 = y`, is this what is happening?
1) There exists
```
NameResolutionContext {
  owner_id: Some(Box0),
  quantifiers: {Q0},
  cte: {"T1" -> Box1},
  parent_context: None,
  sibling_context: None,
  is_lateral: false
}
```
2) There also exists
```
NameResolutionContext {
  owner_id: Some(Box2),
  quantifiers: {Q1, Q2, Q3},
  cte: {"T2" -> Box3, "T3" -> Box5, "T4" -> Box6},
  parent_context: Some(<context mentioned in (1)>),
  sibling_context: None,
  is_lateral: false
}
```
3) After everything in Box2 is resolved, the `NameResolutionContext` mentioned in (2) is destroyed, and Q1 and Q4 are added to the quantifiers of the `NameResolutionContext` mentioned in (1).

Reply:
Almost. The only difference is that context 1 is not the parent context of context 2, but its sibling context, only visible for lateral join operands.

Yes, the quantifiers in the context of a subjoin get merged into the context of the parent join when it goes out of scope. An example can be seen in:

https://github.com/asenac/rust-sql/blob/bd877a189f50f5f6d8b0e07947c15b2a4409b075/src/query_model/model_generator.rs#L655